hadoop-ozone/ozonefs/src/main/java/org/apache/hadoop/fs/ozone/BasicRootedOzoneClientAdapterImpl.java

```diff
@@ -0,0 +1,694 @@
+      }
+
+      String replicationTypeConf =
+          conf.get(OzoneConfigKeys.OZONE_REPLICATION_TYPE,
+              OzoneConfigKeys.OZONE_REPLICATION_TYPE_DEFAULT);
+
+      int replicationCountConf = conf.getInt(OzoneConfigKeys.OZONE_REPLICATION,
+          OzoneConfigKeys.OZONE_REPLICATION_DEFAULT);
+
+      if (OmUtils.isOmHAServiceId(conf, omHost)) {
+        // omHost is listed as one of the service ids in the config,
+        // thus we should treat omHost as omServiceId
+        this.ozoneClient =
+            OzoneClientFactory.getRpcClient(omHost, conf);
+      } else if (StringUtils.isNotEmpty(omHost) && omPort != -1) {
+        this.ozoneClient =
+            OzoneClientFactory.getRpcClient(omHost, omPort, conf);
+      } else {
+        this.ozoneClient =
+            OzoneClientFactory.getRpcClient(conf);
+      }
+      objectStore = ozoneClient.getObjectStore();
+      proxy = objectStore.getClientProxy();
+      this.replicationType = ReplicationType.valueOf(replicationTypeConf);
+      this.replicationFactor = ReplicationFactor.valueOf(replicationCountConf);
+      this.configuredDnPort = conf.getInt(
+          OzoneConfigKeys.DFS_CONTAINER_IPC_PORT,
+          OzoneConfigKeys.DFS_CONTAINER_IPC_PORT_DEFAULT);
+    } finally {
+      Thread.currentThread().setContextClassLoader(contextClassLoader);
+    }
+  }
+
+  OzoneBucket getBucket(OFSPath ofsPath, boolean createIfNotExist)
+      throws IOException {
+
+    return getBucket(ofsPath.getVolumeName(), ofsPath.getBucketName(),
+        createIfNotExist);
+  }
+
+  /**
+   * Get OzoneBucket object to operate in.
+   * Optionally create volume and bucket if not found.
+   *
+   * @param createIfNotExist Set this to true if the caller is a write operation
+   *                         in order to create the volume and bucket.
+   * @throws IOException Exceptions other than OMException with result code
+   *                     VOLUME_NOT_FOUND or BUCKET_NOT_FOUND.
+   */
+  private OzoneBucket getBucket(String volumeStr, String bucketStr,
+      boolean createIfNotExist) throws IOException {
+
+    OzoneBucket bucket;
+    try {
+      bucket = proxy.getBucketDetails(volumeStr, bucketStr);
+    } catch (OMException ex) {
+      if (createIfNotExist) {
+        // Note: getBucketDetails always throws BUCKET_NOT_FOUND, even if
+        // the volume doesn't exist.
+        if (ex.getResult().equals(BUCKET_NOT_FOUND)) {
+          OzoneVolume volume;
+          try {
+            volume = proxy.getVolumeDetails(volumeStr);
+          } catch (OMException getVolEx) {
+            if (getVolEx.getResult().equals(VOLUME_NOT_FOUND)) {
+              // Volume doesn't exist. Create it
+              try {
+                objectStore.createVolume(volumeStr);
+              } catch (OMException newVolEx) {
+                // Ignore the case where another client created the volume
+                if (!newVolEx.getResult().equals(VOLUME_ALREADY_EXISTS)) {
```

Comment:
Even the volume creation succeed, the bucket creation might fail if the ACLs are not set properly when security and acls are enabled. With o3fs, this is OK because the volume/bucket are provisioned and acls setup before it is mounted. With ofs, this may not work without ACL changes.

Reply:
As we discussed, currently in Ozone `createVolume` will set owner to current user when `volArg` owner is null: https://github.com/smengcl/hadoop-ozone/blob/046a06f02783da516179ee8d8d1bed862d22f78d/hadoop-ozone/client/src/main/java/org/apache/hadoop/ozone/client/rpc/RpcClient.java#L268

Can we resolve this one?